libfoundation/src/foundation-java-private.cpp

```diff
@@ -0,0 +1,1353 @@
+    
+    return className;
+}
+
+static bool __JavaJNIInstanceMethodResult(jobject p_instance, jmethodID p_method_id, jvalue *p_params, int p_return_type, void *r_result)
+{
+    MCJavaDoAttachCurrentThread();
+    auto t_return_type = static_cast<MCJavaType>(p_return_type);
+    
+    switch (t_return_type)
+    {
+        case kMCJavaTypeBoolean:
+        {
+            jboolean t_result =
+                s_env -> CallBooleanMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jboolean *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeByte:
+        {
+            jbyte t_result =
+                s_env -> CallByteMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jbyte *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeChar:
+        {
+            jchar t_result =
+                s_env -> CallCharMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jchar *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeShort:
+        {
+            jshort t_result =
+                s_env -> CallShortMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jshort *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeInt:
+        {
+            jint t_result =
+                s_env -> CallIntMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jint *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeLong:
+        {
+            jlong t_result =
+                s_env -> CallLongMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jlong *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeFloat:
+        {
+            jfloat t_result =
+                s_env -> CallFloatMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jfloat *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeDouble:
+        {
+            jdouble t_result =
+                s_env -> CallDoubleMethodA(p_instance, p_method_id, p_params);
+            *(static_cast<jdouble *>(r_result)) = t_result;
+            break;
+        }
+        case kMCJavaTypeObject:
+        case kMCJavaTypeArray:
+        {
+            jobject t_result =
+                s_env -> CallObjectMethodA(p_instance, p_method_id, p_params);
+            
+            MCJavaObjectRef t_result_value;
+            if (!MCJavaObjectCreateGlobalRef(t_result, t_result_value))
+                return false;
+            *(static_cast<MCJavaObjectRef *>(r_result)) = t_result_value;
+            break;
+        }
+        case kMCJavaTypeVoid:
+            s_env -> CallVoidMethodA(p_instance, p_method_id, p_params);
+            break;
+        default:
+            MCUnreachableReturn(false);
+    }
+    return true;
+}
```

Comment:
As we've discussed, these big `switch` could be simplified (and compressed!) with some suitable C++11 magic.

In the short term, it would be nice to get compilation warnings if the `kMCJavaType` enumeration changes.  Because there's a `default` label in the `switch`, we won't get that.

I suggest using this form:

```c++
switch (t_return_type)
{
    case kMCJavaTypeBoolean:
    {
        jboolean t_result =
            s_env -> CallBooleanMethodA(p_instance, p_method_id, p_params);
        *(static_cast<jboolean *>(r_result)) = t_result;
        return true;
    }

    /* ... other cases here */

    /* no default label */
}
MCUnreachableReturn(false);
```